content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+}
+```
+
+Csmith conservatively analyzes the *effect* of every expression. An effect contains *may-read* and *may-write* locations of the expression. The same location cannot be both read and write or written twice except for assignment. This is done incrementally. For newly generated code, a check is performed to decide abandon or keep the code. For example, when generating  `p + funcMayModifyP()`, Csmith will abandon `funcMayModifyP()` .
+
+### Array Safety
+
+Out of boundary safety issues are always there for arrays.
+
+```c++
+void outOfBound(){
+    int a[2];
+    a[2] = 10; // out of boundary
+}
+```
+
+The example I give is simple to avoid, but when the index of `a` is a variable, it is hard to tell whether it is in-bounds or not. Csmith only generates index variables of `for` loops counter and ensures the `for` loop never exceeds the boundary. For arbitrary index variables, Csmith applies the modulo operator. If both techniques do not work—for example, when array length increases—Csmith emits explicit checks against array lengths.
+
+### Initializer Safety
+
+Csmith initializes variables right after declaration and bans `goto` statement to ensure the execution is in order.
+
+**Question:** What kind of cases can be omitted due to the current design of program generation and safety constraints? 
+
+### Global Safety
+
+Because Csmith generates the code incrementally, newly generated code can threaten the old code.
+
+```c++
+incrementallyGeneratedUnsafeProgram(){}
+    int *p = &i;
+    while (...) {
+        *p = 3; 
+        p = 0; // unsafe because of the back-edge
+    }
+}
+```
+
+Csmith performs checking at each newly generated line except for loops. Loops are checked at the end when the back-edge is logically created. If unsafe lines appear, Csmith deletes line by line until the safety requirements are satisfied.
+
+## Design Tradeoff
+
+### Allow Implementation-defined Behavior
+
+Implementation-defined behavior is equivalent to [unspecified behavior](<https://en.wikipedia.org/wiki/Unspecified_behavior>), which may vary on different implementations of a programming language. Csmith designers believe it is unrealistic to "retain a single interpretation across all possible choices of implementation-defined behaviors". They allow Csmith to give different outputs when compilers has implementation-defined behavior at 
+
+1. the widths and representations of integers
+2. behavior when casting to a signed integer type when the value cannot be represented in an object of the target type
+3. results of bitwise operations on signed integers. 
+
+There are roughly three kinds of compiler targets:
+
+| Targets               | `int` bit length | `long` bit length |
+| --------------------- | ---------------- | ----------------- |
+| x86-64                | 32               | 64                |
+| x86, ARM, and PowerPC | 32               | 32                |
+| MSP430 and AVR        | 16               | 16                |
```

Comment:
Probably worth noting that ARMv8 is like x86-64 in this regard, but it didn't exist in 2011.